handlers/weniwebchat/weniwebchat.go

```diff
@@ -0,0 +1,259 @@
+package weniwebchat
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/handlers"
+	"github.com/nyaruka/courier/utils"
+	"github.com/nyaruka/gocommon/urns"
+)
+
+func init() {
+	courier.RegisterHandler(newHandler())
+}
+
+type handler struct {
+	handlers.BaseHandler
+}
+
+func newHandler() courier.ChannelHandler {
+	return &handler{handlers.NewBaseHandler(courier.ChannelType("WWC"), "Weni Web Chat")}
+}
+
+// Initialize is called by the engine once everything is loaded
+func (h *handler) Initialize(s courier.Server) error {
+	h.SetServer(s)
+	s.AddHandlerRoute(h, http.MethodPost, "receive", h.receiveMsg)
+	return nil
+}
+
+type miPayload struct {
+	Type    string    `json:"type"           validate:"required"`
+	From    string    `json:"from,omitempty" validate:"required"`
+	Message miMessage `json:"message"`
+}
+
+type miMessage struct {
+	Type      string `json:"type"          validate:"required"`
+	TimeStamp string `json:"timestamp"     validate:"required"`
+	Text      string `json:"text,omitempty"`
+	MediaURL  string `json:"media_url,omitempty"`
+	Caption   string `json:"caption,omitempty"`
+	Latitude  string `json:"latitude,omitempty"`
+	Longitude string `json:"longitude,omitempty"`
+}
+
+func (h *handler) receiveMsg(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.Event, error) {
+	payload := &miPayload{}
+	err := handlers.DecodeAndValidateJSON(payload, r)
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+	}
+
+	// check message type
+	if payload.Type != "message" || (payload.Message.Type != "text" && payload.Message.Type != "image" && payload.Message.Type != "video" && payload.Message.Type != "audio" && payload.Message.Type != "file" && payload.Message.Type != "location") {
+		return nil, handlers.WriteAndLogRequestIgnored(ctx, h, channel, w, r, "ignoring request, unknown message type")
+	}
+
+	// check empty content
+	if payload.Message.Text == "" && payload.Message.MediaURL == "" && (payload.Message.Latitude == "" || payload.Message.Longitude == "") {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, errors.New("blank message, media or location"))
+	}
+
+	// build urn
+	urn, err := urns.NewURNFromParts(urns.ExternalScheme, payload.From, "", "")
```

Comment:
Any reason to not define a new URN type? And what is `From` here?

Reply:
waiting for nyaruka/gocommon#51 be approved to proceed